mautic/mautic
# Updated permissions for lead list (segment).

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [4.x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [Yes] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:


Currently, if a non-admin user creates a segment, and attempts to edit it, they are not presented with any edit button, and the UI appears to be breaking.

Instead of fixing the permission issue only, I have updated segment-related permissions to match other entities.
This PR will have full permission support for Segments.

These changes will ensure an administrator can manage roles for the segment permission independent of the Lead permission.

The changelog,
1. Added full permissions for segment to choose from.
2. Migration of the old permissions. All the permissions from the Lead are mapped to the segments.
3. The ListController is now using the grant/deny access based on these new permissions.
4. Added code coverage for the recent permissions changes.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull-down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a role selecting segment-related permissions.
3. Create a mautic user and assign it the above-created role.
5. Using the credentials for the new user, browse the different operations on the segments.
6. Repeat the above steps for combinations of permissions.

Please find the following table containing the operation one can perform on segments,
| Page | URI | Permission |
| --- | --- | --- |
| Index (Listing) | `/s/segments` | All or any permission(s) |
| Create New | `s/segments/new` | Create |
| View | `s/segments/view/<segmentId>` | View own and/or View others |
| Edit | `s/segments/edit/<segmentId>` | Edit own and/or Edit other |
| Clone | `s/segments/edit/<segmentId>` | Create and View own and/or View other |
| Delete | `s/segments/delete/<segmentId>` | Delete own and/or Delete other |
| Batch Delete (This operation is available on listing page.) | `s/segments/batchDelete/?ids=<sids>` | Delete own and/or Delete other |
| Contact grid |`s/segments/view/<segmentId>` | View own or View others |





<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10824"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
